### DISCUSSION TITLE
Forge - Ubuntu 22 vs 24 & AppArmor

### DISCUSSION THREAD

**Discussion by RyanGSampson** [spatie/browsershot#900] [Q&A]
Hello All,

I use Forge to deploy my laravel application.
My production server is running on Ubuntu 22 w/ Browsershot performing perfectly. 

I recently provisioned a new dev server using Ubunti 24 and ran into the issue of AppArmor not allowing Puppeteer to spin up Chromim in a sandbox. 

Disabling sandboxing is not something that I am willing to do.

For now, going back to Ubuntu 22 is my solution. 
Wonding if anyone has experience in bypassing AppArmor in 24?

I tried the below to no avail.  
Thanks for reading. 

sudo nano /etc/apparmor.d/local/usr.local.bin.chromium

profile chromium /snap/bin/chromium flags=(unconfined) {
    userns,
    include if exists <local/chromium>
}

sudo apparmor_parser -r /etc/apparmor.d/usr.local.bin.chromium

**Comment by matticustard** [comment]
The following profile is working for me on Forge with Ubuntu 24.04. Browsershot and its dependencies were installed per the documentation.

Create AppArmor profile and reload:

```sh
export CHROMIUM_BUILD_PATH=/@{HOME}/.cache/puppeteer/**/chrome
cat | sudo tee /etc/apparmor.d/chrome-puppeteer <<EOF
abi <abi/4.0>,
include <tunables/global>

profile chrome $CHROMIUM_BUILD_PATH flags=(unconfined) {
  userns,

  # Site-specific additions and overrides. See local/README for details.
  include if exists <local/chrome>
}
EOF
sudo service apparmor reload  # reload AppArmor profiles to include the new one
```

**Reply by edalzell** [reply]
This unfortunately, did not work for me on a new Forge/Ubuntu 24 server. Wonder what I did wrong.

**Reply by talelmishali** [reply]
> This unfortunately, did not work for me on a new Forge/Ubuntu 24 server. Wonder what I did wrong.

Check my comment bellow, worked for me. Instead of messing with AppArmor configs, use the "safest way" as per the documentation offered by the `chromium` team.

**Comment by talelmishali** [comment]
Following the [official Chromium documentation](https://chromium.googlesource.com/chromium/src/+/main/docs/security/apparmor-userns-restrictions.md), use Google Chrome's SUID sandbox, that worked for me!

1. Install Google Chrome:
```bash
wget -q -O - https://dl.google.com/linux/linux_signing_key.pub | sudo apt-key add -
echo "deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main" | sudo tee /etc/apt/sources.list.d/google-chrome.list
sudo apt update && sudo apt install google-chrome-stable
```

2. Set environment variable system-wide:
```bash
echo 'export CHROME_DEVEL_SANDBOX=/opt/google/chrome/chrome-sandbox' | sudo tee -a /etc/environment
```

3. Configure PHP-FPM to pass the environment variable:
Add this to your PHP-FPM pool config (usually `/etc/php/8.4/fpm/pool.d/www.conf`):
```bash
env[CHROME_DEVEL_SANDBOX] = /opt/google/chrome/chrome-sandbox
```

4. Restart PHP-FPM:
```bash
sudo systemctl restart php8.4-fpm
```

**Reply by talelmishali** [reply]
Exposing the environment variable in PHP-FPM makes it available only to web requests. To have it work with a queue worker, you need to edit the Supervisor configuration and define the environment variable there:
```bash
environment=CHROME_DEVEL_SANDBOX="/opt/google/chrome/chrome-sandbox"
```
